src/Workspaces/Core/Portable/Workspace/Solution/ProjectChanges.cs

```diff
@@ -123,7 +123,13 @@ public IEnumerable<DocumentId> GetChangedDocuments()
             {
                 var newState = _newProject.GetDocumentState(id);
                 var oldState = _oldProject.GetDocumentState(id);
-                if (oldState != null && newState != oldState)
+                // should not directly compare newState != oldState,
+                // as when ParseOptions changes we should not consider as a document changes
+                // https://github.com/dotnet/roslyn/issues/18199
```

Comment:
hrmm... this seems *strange*.  ParseOptions are accessibly off of a document.  so if parse options change then it very much seems the case that the document should be considered changed.  After all, if you compare the solutions, you'll have observably different Document objects...

Reply:
but this is the bug is about, when upgrade a project, all the ParseOption.LanguageVersion changed, but the documents did not change and we should not consider they changed...

Reply:
I think there's a distinction to be drawn here:

1. When new parse options are provided, the documents themselves have changed, since @CyrusNajmabadi is correct: the change can be observed. One Document isn't the same as the other.
2. When it comes to the textual content of the document, they _haven't_ changed. Namely, this code shouldn't be seeing this as a changed document:

https://github.com/dotnet/roslyn/blob/cee5906cbdd4f93c702baea5e98cf33b3a5875bd/src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs#L162-L167

I'd also assume that when it comes to previewing we don't want to show this as a changed file _either_. Do we show it today for this quick fix?

Maybe we really need two APIs:

1. This one that stays as it is.
2. GetDocumentsWithChangedText() which does, well, this new code.

...and then update callers from the old to the new one that wanted the new one all along.

Reply:
jason took the words out of my mouth :)

We need to be especially careful here since ProjectChanges and GetChangedDocuments are public APIs.  if we're going to change the behavior here (which i would not recommend), then that would have to go through compat council.

however, as jason said the better approach is just "GetChangedDocuments" and something like "GetTextuallyChangedDocuments"

Reply:
Also, it's still not clear to me why anything is crashing.  Even if we say "the file has changed" just when parse options change... why is that causing a crash?